File: go/text/gift_suggestions.go
package text

import (
	"math/rand"
	"sort"
	"strings"
)

// Gift holds a gift with category, price range, and compatible traits.
type Gift struct {
	Name        string
	Category    string
	PriceRange  string // "budget", "mid", "premium"
	SuitableFor []string
}

// GiftSuggester is an attribute-based gift recommendation engine.
type GiftSuggester struct {
	gifts []Gift
}

// NewGiftSuggester creates a suggester pre-loaded with default gifts.
func NewGiftSuggester() *GiftSuggester {
	gs := &GiftSuggester{}
	gs.loadDefaults()
	return gs
}

func (gs *GiftSuggester) loadDefaults() {
	gs.gifts = []Gift{
		// Tech
		{"Mechanical Keyboard", "tech", "mid", []string{"techie", "creative"}},
		{"Raspberry Pi Kit", "tech", "budget", []string{"techie", "creative"}},
		{"Noise-Cancelling Headphones", "tech", "premium", []string{"techie", "music-lover"}},
		{"Smart Home Starter Kit", "tech", "mid", []string{"techie"}},
		// Books
		{"Leather-Bound Journal", "books", "mid", []string{"bookworm", "creative"}},
		{"Complete Tolkien Collection", "books", "premium", []string{"bookworm", "adventurer"}},
		{"Pocket Poetry Anthology", "books", "budget", []string{"bookworm", "creative"}},
		{"Cookbook: World Cuisines", "books", "mid", []string{"bookworm", "foodie"}},
		// Outdoor
		{"Hammock", "outdoor", "budget", []string{"adventurer"}},
		{"Hiking Backpack", "outdoor", "mid", []string{"adventurer"}},
		{"Camping Cookset", "outdoor", "mid", []string{"adventurer", "foodie"}},
		{"Trail Running Shoes", "outdoor", "premium", []string{"adventurer"}},
		// Cooking
		{"Cast Iron Skillet", "cooking", "budget", []string{"foodie"}},
		{"Spice Collection Box", "cooking", "mid", []string{"foodie", "adventurer"}},
		{"Chef's Knife Set", "cooking", "premium", []string{"foodie"}},
		{"Pasta Maker", "cooking", "mid", []string{"foodie", "creative"}},
		// Music
		{"Vinyl Record Starter Pack", "music", "budget", []string{"music-lover"}},
		{"Concert Tickets", "music", "mid", []string{"music-lover", "adventurer"}},
		{"Turntable", "music", "premium", []string{"music-lover"}},
		{"MIDI Controller", "music", "mid", []string{"music-lover", "techie", "creative"}},
		// Art
		{"Watercolor Set", "art", "budget", []string{"creative"}},
		{"Drawing Tablet", "art", "mid", []string{"creative", "techie"}},
		{"Museum Membership", "art", "mid", []string{"creative", "bookworm"}},
		{"Oil Paint Master Set", "art", "premium", []string{"creative"}},
	}
}

// AddGift adds a gift to the catalog.
func (gs *GiftSuggester) AddGift(name, category, priceRange string, suitableFor []string) {
	gs.gifts = append(gs.gifts, Gift{
		Name:        name,
		Category:    category,
		PriceRange:  priceRange,
		SuitableFor: suitableFor,
	})
}

// scoredGift pairs a relevance score with a gift for sorting.
type scoredGift struct {
	score int
	gift  Gift
}

// Suggest returns gifts matching the given traits, sorted by relevance (descending).
// If budget is non-empty, only gifts in that price range are included.
func (gs *GiftSuggester) Suggest(traits []string, budget string) []Gift {
	traitSet := make(map[string]bool)
	for _, t := range traits {
		traitSet[strings.ToLower(t)] = true
	}

	var scored []scoredGift
	for _, g := range gs.gifts {
		if budget != "" && g.PriceRange != budget {
			continue
		}
		matches := 0
		for _, s := range g.SuitableFor {
			if traitSet[strings.ToLower(s)] {
				matches++
			}
		}
		if matches > 0 {
			scored = append(scored, scoredGift{score: matches, gift: g})
		}
	}

	sort.Slice(scored, func(i, j int) bool {
		return scored[i].score > scored[j].score
	})

	result := make([]Gift, len(scored))
	for i, sg := range scored {
		result[i] = sg.gift
	}
	return result
}

// RandomSuggestion returns a random gift from the catalog.
func (gs *GiftSuggester) RandomSuggestion() (Gift, bool) {
	if len(gs.gifts) == 0 {
		return Gift{}, false
	}
	return gs.gifts[rand.Intn(len(gs.gifts))], true
}

// SuggestByCategory returns all gifts in a given category.
func (gs *GiftSuggester) SuggestByCategory(category string) []Gift {
	cat := strings.ToLower(category)
	var result []Gift
	for _, g := range gs.gifts {
		if strings.ToLower(g.Category) == cat {
			result = append(result, g)
		}
	}
	return result
}

File: rust/text/gift_suggestions.rs
//! Random Gift Suggestions — attribute-based gift recommendation.

use std::collections::HashSet;

/// A gift with category, price range, and compatible traits.
#[derive(Debug, Clone)]
pub struct Gift {
    pub name: String,
    pub category: String,
    pub price_range: String, // "budget", "mid", "premium"
    pub suitable_for: Vec<String>,
}

impl std::fmt::Display for Gift {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} [{}] ({}) — suits: {}",
            self.name,
            self.category,
            self.price_range,
            self.suitable_for.join(", ")
        )
    }
}

/// Attribute-based gift recommendation engine.
pub struct GiftSuggester {
    gifts: Vec<Gift>,
}

impl GiftSuggester {
    /// Create a new suggester pre-loaded with default gifts.
    pub fn new() -> Self {
        let mut gs = Self { gifts: Vec::new() };
        gs.load_defaults();
        gs
    }

    fn load_defaults(&mut self) {
        let defaults = vec![
            // Tech
            ("Mechanical Keyboard", "tech", "mid", vec!["techie", "creative"]),
            ("Raspberry Pi Kit", "tech", "budget", vec!["techie", "creative"]),
            ("Noise-Cancelling Headphones", "tech", "premium", vec!["techie", "music-lover"]),
            ("Smart Home Starter Kit", "tech", "mid", vec!["techie"]),
            // Books
            ("Leather-Bound Journal", "books", "mid", vec!["bookworm", "creative"]),
            ("Complete Tolkien Collection", "books", "premium", vec!["bookworm", "adventurer"]),
            ("Pocket Poetry Anthology", "books", "budget", vec!["bookworm", "creative"]),
            ("Cookbook: World Cuisines", "books", "mid", vec!["bookworm", "foodie"]),
            // Outdoor
            ("Hammock", "outdoor", "budget", vec!["adventurer"]),
            ("Hiking Backpack", "outdoor", "mid", vec!["adventurer"]),
            ("Camping Cookset", "outdoor", "mid", vec!["adventurer", "foodie"]),
            ("Trail Running Shoes", "outdoor", "premium", vec!["adventurer"]),
            // Cooking
            ("Cast Iron Skillet", "cooking", "budget", vec!["foodie"]),
            ("Spice Collection Box", "cooking", "mid", vec!["foodie", "adventurer"]),
            ("Chef's Knife Set", "cooking", "premium", vec!["foodie"]),
            ("Pasta Maker", "cooking", "mid", vec!["foodie", "creative"]),
            // Music
            ("Vinyl Record Starter Pack", "music", "budget", vec!["music-lover"]),
            ("Concert Tickets", "music", "mid", vec!["music-lover", "adventurer"]),
            ("MIDI Controller", "music", "mid", vec!["music-lover", "techie", "creative"]),
            ("Turntable", "music", "premium", vec!["music-lover"]),
            // Art
            ("Watercolor Set", "art", "budget", vec!["creative"]),
            ("Drawing Tablet", "art", "mid", vec!["creative", "techie"]),
            ("Museum Membership", "art", "mid", vec!["creative", "bookworm"]),
            ("Oil Paint Master Set", "art", "premium", vec!["creative"]),
        ];
        for (name, cat, price, traits) in defaults {
            self.gifts.push(Gift {
                name: name.to_string(),
                category: cat.to_string(),
                price_range: price.to_string(),
                suitable_for: traits.into_iter().map(String::from).collect(),
            });
        }
    }

    /// Add a gift to the catalog.
    pub fn add_gift(&mut self, name: &str, category: &str, price_range: &str, suitable_for: &[&str]) {
        self.gifts.push(Gift {
            name: name.to_string(),
            category: category.to_string(),
            price_range: price_range.to_string(),
            suitable_for: suitable_for.iter().map(|s| s.to_string()).collect(),
        });
    }

    /// Suggest gifts matching traits, sorted by relevance (descending).
    /// If budget is Some, only gifts in that price range are returned.
    pub fn suggest(&self, traits: &[&str], budget: Option<&str>) -> Vec<&Gift> {
        let trait_set: HashSet<String> = traits.iter().map(|t| t.to_lowercase()).collect();
        let mut scored: Vec<(usize, &Gift)> = self
            .gifts
            .iter()
            .filter(|g| budget.map_or(true, |b| g.price_range == b))
            .filter_map(|g| {
                let gift_traits: HashSet<String> =
                    g.suitable_for.iter().map(|s| s.to_lowercase()).collect();
                let matches = trait_set.intersection(&gift_traits).count();
                if matches > 0 { Some((matches, g)) } else { None }
            })
            .collect();
        scored.sort_by(|a, b| b.0.cmp(&a.0));
        scored.into_iter().map(|(_, g)| g).collect()
    }

    /// Return a random gift from the catalog.
    pub fn random_suggestion(&self) -> Option<&Gift> {
        if self.gifts.is_empty() {
            return None;
        }
        use std::time::SystemTime;
        let seed = SystemTime::now()
            .duration_since(SystemTime::UNIX_EPOCH)
            .unwrap_or_default()
            .subsec_nanos() as usize;
        Some(&self.gifts[seed % self.gifts.len()])
    }

    /// Return all gifts in a given category.
    pub fn suggest_by_category(&self, category: &str) -> Vec<&Gift> {
        let cat = category.to_lowercase();
        self.gifts.iter().filter(|g| g.category.to_lowercase() == cat).collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_catalog_size() {
        let gs = GiftSuggester::new();
        assert_eq!(gs.gifts.len(), 24);
    }

    #[test]
    fn test_suggest_single_trait() {
        let gs = GiftSuggester::new();
        let results = gs.suggest(&["foodie"], None);
        assert!(!results.is_empty());
        for gift in &results {
            assert!(gift.suitable_for.iter().any(|t| t == "foodie"));
        }
    }

    #[test]
    fn test_suggest_with_budget() {
        let gs = GiftSuggester::new();
        let results = gs.suggest(&["techie"], Some("budget"));
        for gift in &results {
            assert_eq!(gift.price_range, "budget");
        }
    }

    #[test]
    fn test_suggest_relevance_order() {
        let gs = GiftSuggester::new();
        let results = gs.suggest(&["techie", "music-lover", "creative"], None);
        assert!(!results.is_empty());
        assert_eq!(results[0].name, "MIDI Controller");
    }

    #[test]
    fn test_suggest_by_category() {
        let gs = GiftSuggester::new();
        let cooking = gs.suggest_by_category("cooking");
        assert_eq!(cooking.len(), 4);
        for gift in &cooking {
            assert_eq!(gift.category, "cooking");
        }
    }

    #[test]
    fn test_add_gift() {
        let mut gs = GiftSuggester::new();
        gs.add_gift("Custom Widget", "tech", "premium", &["techie"]);
        assert_eq!(gs.gifts.len(), 25);
        let results = gs.suggest_by_category("tech");
        assert!(results.iter().any(|g| g.name == "Custom Widget"));
    }

    #[test]
    fn test_random_suggestion() {
        let gs = GiftSuggester::new();
        assert!(gs.random_suggestion().is_some());
    }

    #[test]
    fn test_no_matches() {
        let gs = GiftSuggester::new();
        let results = gs.suggest(&["nonexistent-trait"], None);
        assert!(results.is_empty());
    }
}
